File: lib/game/systems/save_system.dart
import 'package:shared_preferences/shared_preferences.dart';

/// Thin wrapper around SharedPreferences for the game's persistent state:
/// best distance, total Sparks (currency), unlocked/selected character.
class SaveSystem {
  static const _kBestDistance = 'gati.best_distance';
  static const _kTotalSparks = 'gati.total_sparks';
  static const _kUnlocked = 'gati.unlocked_characters';
  static const _kSelected = 'gati.selected_character';

  static SharedPreferences? _prefs;

  static Future<void> init() async {
    _prefs ??= await SharedPreferences.getInstance();
  }

  static double get bestDistance => _prefs?.getDouble(_kBestDistance) ?? 0;
  static Future<void> setBestDistanceIfHigher(double meters) async {
    if (meters > bestDistance) {
      await _prefs?.setDouble(_kBestDistance, meters);
    }
  }

  static int get totalSparks => _prefs?.getInt(_kTotalSparks) ?? 0;
  static Future<void> addSparks(int amount) async {
    await _prefs?.setInt(_kTotalSparks, totalSparks + amount);
  }

  static Future<bool> spendSparks(int amount) async {
    if (totalSparks < amount) return false;
    await _prefs?.setInt(_kTotalSparks, totalSparks - amount);
    return true;
  }

  static List<String> get unlockedCharacterIds =>
      _prefs?.getStringList(_kUnlocked) ?? const ['school_kid'];

  static Future<void> unlockCharacter(String id) async {
    final list = {...unlockedCharacterIds, id}.toList();
    await _prefs?.setStringList(_kUnlocked, list);
  }

  static String get selectedCharacterId => _prefs?.getString(_kSelected) ?? 'school_kid';
  static Future<void> selectCharacter(String id) async {
    await _prefs?.setString(_kSelected, id);
  }
}

File: unity/Assets/Scripts/Core/SaveSystem.cs
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

namespace Gati.Core
{
    /// <summary>Thin wrapper around PlayerPrefs for persistent game state.</summary>
    public static class SaveSystem
    {
        const string KeyBestDistance = "gati.best_distance";
        const string KeyTotalSparks = "gati.total_sparks";
        const string KeyUnlocked = "gati.unlocked_characters";
        const string KeySelected = "gati.selected_character";

        public static float BestDistance => PlayerPrefs.GetFloat(KeyBestDistance, 0f);

        public static void SetBestDistanceIfHigher(float meters)
        {
            if (meters > BestDistance)
            {
                PlayerPrefs.SetFloat(KeyBestDistance, meters);
                PlayerPrefs.Save();
            }
        }

        public static int TotalSparks => PlayerPrefs.GetInt(KeyTotalSparks, 0);

        public static void AddSparks(int amount)
        {
            PlayerPrefs.SetInt(KeyTotalSparks, TotalSparks + amount);
            PlayerPrefs.Save();
        }

        public static bool SpendSparks(int amount)
        {
            if (TotalSparks < amount) return false;
            PlayerPrefs.SetInt(KeyTotalSparks, TotalSparks - amount);
            PlayerPrefs.Save();
            return true;
        }

        public static List<string> UnlockedCharacterIds
        {
            get
            {
                var raw = PlayerPrefs.GetString(KeyUnlocked, "school_kid");
                return raw.Split(',').Where(s => s.Length > 0).ToList();
            }
        }

        public static void UnlockCharacter(string id)
        {
            var set = new HashSet<string>(UnlockedCharacterIds) { id };
            PlayerPrefs.SetString(KeyUnlocked, string.Join(",", set));
            PlayerPrefs.Save();
        }

        public static string SelectedCharacterId => PlayerPrefs.GetString(KeySelected, "school_kid");

        public static void SelectCharacter(string id)
        {
            PlayerPrefs.SetString(KeySelected, id);
            PlayerPrefs.Save();
        }
    }
}
